src/codemodder/codetf/v3/codetf.py

```diff
@@ -0,0 +1,181 @@
+from __future__ import annotations
+
+from enum import Enum
+from typing import Optional
+
+from pydantic import BaseModel, model_validator
+
+from ..common import CaseInsensitiveEnum, CodeTFWriter
+
+
+class Run(BaseModel):
+    """Metadata about the analysis run that produced the results"""
+
+    vendor: str
+    tool: str
+    version: str
+    # Optional free-form metadata about the project being analyzed
+    # e.g. project name, directory, commit SHA, etc.
+    projectMetadata: Optional[str] = None
+    # Analysis duration in milliseconds
+    elapsed: Optional[int] = None
+    # Optional free-form metadata about the inputs used for the analysis
+    # e.g. command line, environment variables, etc.
+    inputMetadata: Optional[dict] = None
+    # Optional free-form metadata about the analysis itself
+    # e.g. timeouts, memory usage, etc.
+    analysisMetadata: Optional[dict] = None
+
+
+class FixStatusType(str, Enum):
+    """Status of a fix"""
+
+    fixed = "fixed"
+    skipped = "skipped"
+    failed = "failed"
+    wontfix = "wontfix"
+
+
+class FixStatus(BaseModel):
+    """Metadata describing fix outcome"""
+
+    status: FixStatus
+    reason: Optional[str]
+    details: Optional[str]
+
+
+class Rule(BaseModel):
+    id: str
+    name: str
+    url: Optional[str] = None
+
+
+class Finding(BaseModel):
+    id: str
+    rule: Optional[Rule] = None
+
+
+class Action(CaseInsensitiveEnum):
+    ADD = "add"
+    REMOVE = "remove"
+
+
+class PackageResult(CaseInsensitiveEnum):
+    COMPLETED = "completed"
+    FAILED = "failed"
+    SKIPPED = "skipped"
+
+
+class DiffSide(CaseInsensitiveEnum):
+    LEFT = "left"
+    RIGHT = "right"
+
+
+class PackageAction(BaseModel):
+    action: Action
+    result: PackageResult
```

Comment:
I'm not sure I ever fully understood the rationale for the `result` field: it might be worth discussing whether it's still necessary.

Reply:
I think associating a change to a dependency can allow downstream users to tell users add a dependency (like if it failed), or that this increases reliance on a dependency, etc.